File: Cpp/509.斐波那契数.cpp
/*
 * @lc app=leetcode.cn id=509 lang=cpp
 *
 * [509] 斐波那契数
 */

#include <iostream>

using namespace std;

// @lc code=start
class Solution {
public:
    int fib(int n) {
        if (n == 0) return 0;
        int a = 0, b = 1;
        for (int i = 1; i < n; i++)
        {
            b = a + b;
            a = b - a;
        }
        return b;
    }
};
// @lc code=end

int main(int argc, const char * argv[]) {

    assert(0 == Solution().fib(0));
    assert(1 == Solution().fib(1));
    assert(1 == Solution().fib(2));
    assert(2 == Solution().fib(3));
    assert(3 == Solution().fib(4));
    assert(5 == Solution().fib(5));
    assert(8 == Solution().fib(6));
    assert(13 == Solution().fib(7));
    assert(21 == Solution().fib(8));

    cout << "OK~" << endl;

    return 0;
}

File: Swift/509.斐波那契数.swift
/*
 * @lc app=leetcode.cn id=509 lang=swift
 *
 * [509] 斐波那契数
 */

// @lc code=start
class Solution {
    func fib(_ n: Int) -> Int {
        if n == 0 {
            return 0
        }
        var f0 = 0
        var f1 = 1
        for _ in 1..<n {
            let tmp = f1
            f1 = f1 + f0
            f0 = tmp
        }
        return f1
    }
}
// @lc code=end

func main() {
    var n: Int
    var res: Int
    
    /**
     输入：2
     输出：1
     解释：F(2) = F(1) + F(0) = 1 + 0 = 1
     */
    n = 2
    res = 1
    assert(res == Solution().fib(n))
    
    /**
     输入：3
     输出：2
     解释：F(3) = F(2) + F(1) = 1 + 1 = 2
     */
    n = 3
    res = 2
    assert(res == Solution().fib(n))
    
    /**
     输入：4
     输出：3
     解释：F(4) = F(3) + F(2) = 2 + 1 = 3
     */
    n = 4
    res = 3
    assert(res == Solution().fib(n))
}
